### PR TITLE
[TASK-2.1] Add rest service with crud operations

### PR DESCRIPTION
**Pull request by angulawk**

### CODE REVIEW

File: dist/task4/task4.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+"use strict";
```
Comment (dragonflypl) [inline dist/task4/task4.js]: please add dist folder to gitignore

File: package.json [hunk 2]
```diff
@@ -6,9 +6,10 @@
   "scripts": {
     "test": "echo \"Error: no test specified\" && exit 1",
     "babel-node": "babel-node --presets=/*a*/",
-    "task1": "nodemon ./task1.js localhost 8080",
-    "task2": "nodemon ./task2.js localhost 8080",
-    "task3": "nodemon --exec babel-node -- ./task3.js"
+    "task1": "nodemon ./src/task1.js localhost 8080",
+    "task2": "nodemon ./src/task2.js localhost 8080",
+    "task3": "nodemon --exec babel-node -- ./src/task3.js",
+    "task4": "tsc && nodemon --exec babel-node -- ./dist/task4/task4.js"
```
Comment (dragonflypl) [inline package.json]: why you're running file from dist folder?

File: src/utils/getAutoSuggestUsers.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { getUsers } from "./users";
+import {
+  User
+} from "../task4/typings";
+
+const getAutoSuggestUsers: User[]|any = (loginSubstring: string, limit: number) => {
```
Comment (dragonflypl) [inline src/utils/getAutoSuggestUsers.ts]: why `any` ? It always returns `User[]`

File: src/utils/getAutoSuggestUsers.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+  const filteredArray: User[] = getUsers().filter((item: User) => item.login.includes(loginSubstring));
+  filteredArray
+    .sort(( a: User, b: User ) => {
+      return (a.login > b.login) ? 1 : ((b.login > a.login) ? -1 : 0)
+    })
+    .splice(limit, filteredArray.length - 1);
```
Comment (dragonflypl) [inline src/utils/getAutoSuggestUsers.ts]: `.slice(0, limit)` looks more readable

File: package.json (hunk 2 above)
Comment (angulawk) [reply package.json]: Because my typescript files are compiled from src folder to disc folder

File: src/task4/task4.ts [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+});
+
+app.put("/user/:id", (req: Request, res: Response, next: NextFunction) => {
+  const updatedUsers: User[] = getUsers().map((user: User) => user.id === req.params.id ?
+    { id: user.id,
+      ...req.body
+    } : user);
+  updateUsers(updatedUsers);
+
+  res.json(getUsers());
```
Comment (dragonflypl) [inline src/task4/task4.ts]: read about what put/post/delete should return in proper RESTful API

File: src/task4/task4.ts [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+});
+
+app.get("/auto-suggest/:loginSubstring", (req: Request, res: Response, next: NextFunction) => {
+  const autoSuggestUsers: User[] = getAutoSuggestUsers(req.params.loginSubstring, req.query.limit);
+
+  res.json(autoSuggestUsers);
+  next();
+});
+
+app.delete("/user/:id", (req: Request, res: Response, next: NextFunction) => {
```
Comment (dragonflypl) [inline src/task4/task4.ts]: Move implemetation details of mapping/filtering etc into the users code, e.g.

```
app.delete("/user/:id", (req: Request, res: Response, next: NextFunction) => {
  const users = deleteUser(req.params.id);
  res.json(users );
  next();
});
```

**Pull request closed without merging**